Enabling more ruff lint rules

martinhoyer opened the pull request:
There are too many code changes required to make the checks pass and I don't really expect all of them to be accepted. It should be easy to drop whatever is deemed unnecessary, too strict or just not suitable for this project - probably better to review on per-commit basis.
I've tried to run the default set of tests, but my machine ran out of disk space :clown_face:, not going to claim I didn't break anything.

 - [Polarion report UTC timezone](https://github.com/teemtee/tmt/commit/eb8700945bb5dbdf464e7d41f11a8a7f7075ea26)
Not sure if make sense, but from what I understand the time stamp is being used to generate Title of the test run. Without timezone, I assume the datetime is naive, dependent on the OS settings/location of where the tmt is being run, resulting in potential mismatches, no?

 - Would it be beneficial to add a couple of lines in the contribute doc about ruff? 
Maybe only mention when it's acceptable to add `noqa` (plus to add a comment why it's being added) and/or add a link to [ruff rules page](https://beta.ruff.rs/docs/rules/) to keep it succinct? 
Alternatively, make it more verbose and also mention mypy?

 - Isort change being explained [here](https://github.com/teemtee/tmt/discussions/2003)

 - Can $maintainer(s) check the remaining, commented-out ruff rules in pyproject.toml. I think it would be nice if we could delete those not suitable for this project and keep only those that might get implemented in the future...Or just remove all of them.
 
Notes for some of the disabled checks:

TRY200, B904: Within an `except` clause, raise exceptions with `raise ... from err` or `raise from None.
I've actually tried to implement [PEP-3134](https://peps.python.org/pep-3134/), as it looks like a good practice to me, but ultimately figured it's best to ask about it first.

Example of change in existing code:
```
    except json.decoder.JSONDecodeError as error:
        raise GeneralError(f"Invalid json syntax: {error}")
```
```
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
tmt.utils.GeneralError: Invalid json syntax: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```
changed to: 
```
    except json.decoder.JSONDecodeError as error:
        raise GeneralError("Invalid json syntax") from error
```
```
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
tmt.utils.GeneralError: Invalid json syntax
```

Q: flake8-quotes (Q000 Single quotes found but double quotes preferred) would need changes in basically whole codebase:
> ruff . --select Q000
> Found 4776 errors.

C90: mccabe (complex structure) 

> [tool.ruff.mccabe]
> max-complexity = 14
> 
> $ ruff . --select C90 
> tmt/beakerlib.py:76:9: C901 `__init__` is too complex (17 > 14)
> tmt/beakerlib.py:185:9: C901 `fetch` is too complex (19 > 14)
> tmt/convert.py:191:5: C901 `read_datafile` is too complex (33 > 14)
> tmt/convert.py:380:5: C901 `read` is too complex (41 > 14)
> tmt/convert.py:616:5: C901 `read_nitrate` is too complex (31 > 14)
> tmt/convert.py:758:5: C901 `read_polarion` is too complex (16 > 14)
> tmt/convert.py:863:5: C901 `read_nitrate_case` is too complex (26 > 14)
> tmt/convert.py:1046:5: C901 `relevancy_to_adjust` is too complex (16 > 14)
> tmt/export/__init__.py:295:5: C901 `check_md_file_respects_spec` is too complex (20 > 14)
> tmt/export/nitrate.py:95:5: C901 `convert_manual_to_nitrate` is too complex (16 > 14)
> tmt/export/nitrate.py:358:5: C901 `export_to_nitrate` is too complex (60 > 14)
> tmt/export/polarion.py:150:5: C901 `export_to_polarion` is too complex (41 > 14)
> tmt/options.py:237:5: C901 `create_method_class` is too complex (17 > 14)
> tmt/steps/discover/fmf.py:250:9: C901 `go` is too complex (57 > 14)
> tmt/steps/provision/mrack.py:69:5: C901 `import_and_load_mrack_deps` is too complex (24 > 14)

ANN: flake8-annotations - Some of the checks could compliment mypy.

S: flake8-bandit - Security/vulnerability checks tmt probably doesn't need?

ISC:  ISC001 and ISC003 is actually passing already. Not sure if/how to adapt the code for ISC002 (Implicitly concatenated string literals over multiple lines)

happz commented:
After a brief reading, there are some checks I find arguable, I have questions WRT to some changes, but in general, the majority looks beneficial though, +1 to this effort from me.

And here comes the "but"... I'm afraid it might be way too big for a reasonable review, as a single PR :( A collection of PRs, one for a commit/a group of similar commits (like https://github.com/teemtee/tmt/pull/2014/commits/9ffff4d55b6d4daa1fd580584048937b522beee8 + https://github.com/teemtee/tmt/pull/2014/commits/ed34b66056f368e690984c6c6b6690f4b4933de0) - commits *do* seem to be very nicely rounded - would be easier to digest. Well, at least for me, YMMW. And we could discuss whether we do or do not enable a particular check.

martinhoyer commented on tmt/convert.py:
```diff
@@ -164,7 +161,7 @@ def write_markdown(path: Path, content: Dict[str, str]) -> None:
             echo(style(
                 f"Test case successfully stored into '{path}'.", fg='magenta'))
-    except IOError:
+    except OSError:
```
[pyupgrade](https://github.com/asottile/pyupgrade#oserror-aliases) UP024

happz commented:
> After a brief reading, there are some checks I find arguable, I have questions WRT to some changes, but in general, the majority looks beneficial though, +1 to this effort from me.
> 
> And here comes the "but"... I'm afraid it might be way too big for a reasonable review, as a single PR :( A collection of PRs, one for a commit/a group of similar commits (like [9ffff4d](https://github.com/teemtee/tmt/commit/9ffff4d55b6d4daa1fd580584048937b522beee8) + [ed34b66](https://github.com/teemtee/tmt/commit/ed34b66056f368e690984c6c6b6690f4b4933de0)) - commits _do_ seem to be very nicely rounded - would be easier to digest. Well, at least for me, YMMW. And we could discuss whether we do or do not enable a particular check.

TBH, I'm even scared to start adding comments as I worry it might be very hard to address them e.g. in the middle one of those commits :)

happz commented on tmt/base.py:
```diff
@@ -594,7 +594,7 @@ def web_link(self) -> Optional[str]:
     def _save_context(cls, context: 'tmt.cli.Context') -> None:
         """ Save provided command line context for future use """
-        super(Core, cls)._save_context(context)
+        super()._save_context(context)
```
`super()` should work in class methods too, the simplified call should have the same outcome.

happz commented on pyproject.toml:
```diff
@@ -58,7 +58,7 @@ select = [
 #    "I",  # isort
     "N",  # pep8-naming
-#    "UP",  # pyupgrade
+    "UP",  # pyupgrade
```
Is it possible to control the base Python? The code must stay compatible with Python 3.6.

martinhoyer replied:
Unfortunately ruff uses global Python version settings for all it's check, so while `pyupgrade` supports 3.6, ruff only goes to 3.7.  
I only found one instance of 3.6 compatibility issue and added `noqa` for that line. (subprocess.run universal_newlines)

martinhoyer replied:
> I'd guess it reads either target-version or python_version but both are set to '3.7'. We need to have this as 3.6

Should be target-version in [tool.ruff], which only accepts py37+ see for example https://github.com/charliermarsh/ruff/issues/3826 This is already in the main, the question is whether to enable pyupgrade or not.

martinhoyer replied:
Will disable the check then.  
Now, would it be ok to leave the changes, as they are all 3.6+? Or it's not worth to do code-style changes now, if they're not part of enabling a lint rule?

happz reviewed (commented):
Typo in a commit message: `Fix flake8 C408: unnecessary collecion calls.` => `s/collecion/collection`

happz commented on tests/unit/__init__.py:
```diff
@@ -141,11 +141,7 @@ def _report(message: str) -> None:
-
-{chr(10).join(formatted_fields)}
-"""
+        raise AssertionError(f"\n{message}:\n\n{chr(10).join(formatted_fields)}\n")
```
This is arguable, I'd prefer `assert` to stay. These files are tests, there is no risk for them to be executed with `-O1`. `assert` in tests is IMO perfectly acceptable. Raising an exception makes to code harder to read, also in this example, the formatting is harder to spot.

Also, why just two files? There are plenty of `assert` uses under `tests/unit`, why aren't those reported by ruff?

martinhoyer replied:
From [flake8-bugbear](https://github.com/PyCQA/flake8-bugbear):
"B011: Do not call assert False since python -O removes these calls. Instead callers should raise AssertionError()."

martinhoyer replied:
Fair enough. In addition to B011, it also violates PT015 - "Assertion always fails, replace with `pytest.fail()`".  
noqa for both?

happz replied:
Hmm, `pytest.fail()` actually sounds like a more fitting step to take here, `assert False` is just a less clear way of saying "fail here and now".

happz commented on tests/unit/test_utils.py:
```diff
@@ -609,7 +609,7 @@ def test_run_big(root_logger):
 
 def test_get_distgit_handler():
-    for wrong_remotes in [[], ["blah"]]:
+    for _wrong_remotes in [[], ["blah"]]:
```
This looks like a typo, `_` seems to be the right choice here as `_wrong_remotes` is as unused as `wrong_remotes` was.

happz replied:
Yeah, but, why not replace the name with `_`, getting rid of the unused name completely? Why just prefix it, when it is, as you said, unused? It was good enough for two other loops, so why not this one? :) We don't need `_wrong_remotes` to stick around.

martinhoyer replied:
Well, I don't have a horse in this race and am happy to change it to whatever, though the difference with other loops is that that `_` replaced a single letter character. 
In `for t in threads` I can see we're iterating threads on first glance even without the `t`, while here it might be beneficial to keep the word(s) for better readability.

happz commented on tmt/steps/provision/podman.py:
```diff
@@ -275,18 +275,27 @@ def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecor
+                "--user",
+                metavar="USER",
+                help="User to use for all container operations."),
+            *super().options(how)]
```
I really dislike how the fix is buried under changes that are not related to it :/ Reformatting all options just to put the `super()` into the list, that seems like a way too much noise for such a trivial change :(

martinhoyer replied:
I find it more readable this way, but can revert the formatting. It will pass afaik.

happz replied:
I'm not against the reformatting per se, just against doing it there and under the hood of moving `*super()` into a list literal. Some time ago, the change of the line length was discussed, and while it was increased, tmt code base was not reformatted to make use of the new length, as the patch would introduce quite a bit of noise into `git blame`. This feels like the very same case. Please, feel free to bring it to Tuesday's hacking meeting, but I'd advise against reformatting for the sake of reformatting, especially when it clouds the actual change.

martinhoyer replied:
Reformatting was a bad idea, thanks for keeping me honest. Changed back.

happz commented on examples/plugins/provision.py:
```diff
@@ -30,12 +30,16 @@ def options(cls, how=None):
+                '--switch',
+                is_flag=True,
+                help='Example how to enable something.'),
+            *super().options(how)]
```
Seems to be quite a common behavior, and it does not look like it belongs in this commit which announces addressing of naming issues.

martinhoyer replied:
hmm, looks like I've made a mistake in the `Fix ruff N: pep8-naming errors` commit and added unrelated changes to it. Thanks for catching it.

martinhoyer replied:
Rebased the commits, so hopefully they are correct now.  Keeping the formatting, it was a mistake for changing it, thanks for the review!

happz commented on tmt/utils.py:
```diff
@@ -3547,14 +3546,7 @@ def _patch_plan_schema(schema: Schema, store: SchemaStore) -> None:
-                        }
-                    }
-                ]
+            'oneOf': [*refs, {'type': 'array', 'items': {'anyOf': refs}}]
```
Reformatting feels definitely like a step back, the visual hints WRT structure is lost.

martinhoyer replied:
Keeping the original formatting, at least as much as autopep8 allows.

martinhoyer commented on tmt/base.py:
```diff
@@ -1975,6 +1975,8 @@ def __init__(self,
 
+        if not path:
+            path = Path.cwd()
         self._path = path
```
Nice. Thanks.

martinhoyer commented on tmt/steps/provision/artemis.py:
```diff
@@ -335,7 +335,7 @@ def _create(self) -> None:
         self.info('guestname', self.guestname, 'green')
 
-        with updatable_message(
+        with UpdatableMessage(
```
This would mean renaming the tmt.utils.py `class UpdatableMessage(contextlib.AbstractContextManager)` back to lowercase and adding noqa for N801. Do you want me to do that?

happz replied:
Yes, please. Unless greater powers step in, I for one prefer `updatable_message` (and `retry_session`) - they did pass their review, and I hate the idea of a possible rename if we decide to rewrite them as functions with `yield` :)

martinhoyer replied:
Reverted to `updatable_message` and `retry_session`. Thanks!

happz commented on tmt/base.py:
```diff
@@ -2406,7 +2401,7 @@ def __init__(self,
         # tmt run discover --how fmf --help to create a new workdir.
         super().__init__(context=context, logger=logger)
-        self._workdir_path: WorkdirArgumentType = id_ or True
+        self._workdir_path: WorkdirArgumentType = True
```
This is suspicious. Any idea why `id_` would no longer be an interesting value? It might become `Path` thanks to `id_ = self.config.last_run`.

martinhoyer replied:
Nice catch! I've opened https://github.com/charliermarsh/ruff/issues/4058 and disabling SIM222 for now.

happz commented on tmt/steps/provision/podman.py:
```diff
@@ -315,7 +315,9 @@ def go(self) -> None:
-            for key in PodmanGuestData.keys()
+            # SIM118: Use `{key} in {dict}` instead of `{key} in {dict}.keys()`.
+            # "Type[PodmanGuestData]" has no attribute "__iter__" (not iterable)
+            for key in PodmanGuestData.keys()  # noqa: SIM118
```
Correct, `PodmanGuestData` (and various `*Data` classes) are not dictionaries. They do have `keys/values/items` methods, but are not iterable. Would it help if they become more dict-like & gain `__iter__`?

martinhoyer replied:
Probably not for me to answer...or to say whether such change should be part of this PR for that matter :)

happz replied:
Filed https://github.com/teemtee/tmt/issues/2049

happz commented on tmt/log.py:
```diff
@@ -331,7 +333,7 @@ def filter(self, record: logging.LogRecord) -> bool:
             return True
 
-        return True if details['logger_verbosity_level'] >= message_verbosity_level else False
+        return details['logger_verbosity_level'] >= message_verbosity_level
```
Now this is weird, there's a change down below:

```diff
-            return True if state == 'running' else False
+            return bool(state == "running")
```

Any idea why `bool()` is not applied here as well?

martinhoyer replied:
The `bool()` is there because state is `any`. There is a comment above it.

happz commented on tmt/utils.py:
```diff
@@ -521,6 +521,7 @@ def run(
 
         # Run the command in an interactive mode if requested
         if interactive:
+            # Interactive mode can return non-zero if the last command failed, ignore errors here.
```
Belongs to the commit where `suppress` is added.

martinhoyer replied:
Thanks, fixed.

happz commented:
> * [Polarion report UTC timezone](https://github.com/teemtee/tmt/commit/eb8700945bb5dbdf464e7d41f11a8a7f7075ea26)
>   Not sure if make sense, but from what I understand the time stamp is being used to generate Title of the test run. Without timezone, I assume the datetime is naive, dependent on the OS settings/location of where the tmt is being run, resulting in potential mismatches, no?

Makes sense to me, but I'm no Polarion user. Sticking to UTC seems like a way less prone to misunderstandings.

> * Would it be beneficial to add a couple of lines in the contribute doc about ruff?
>   Maybe only mention when it's acceptable to add `noqa` (plus to add a comment why it's being added) and/or add a link to [ruff rules page](https://beta.ruff.rs/docs/rules/) to keep it succinct?
>   Alternatively, make it more verbose and also mention mypy?

Absolutely, with new linters & influx of waivers, let's communicate "the rules". Especially the comment with check ID & justification needs to be mentioned. I'd maybe add a new section before https://github.com/teemtee/tmt/blob/main/docs/contribute.rst#tests as pre-commit, linters involved, links and rules would consume a bit more than a single line, and they don't fit the tmt-ish nature of the "Tests" chapter. On the other hand, "Tests" does mention unit tests, maybe we could just structure "Tests" a bit - linters, unit tests, testsuite.

> * Can $maintainer(s) check the remaining, commented-out ruff rules in pyproject.toml. I think it would be nice if we could delete those not suitable for this project and keep only those that might get implemented in the future...Or just remove all of them.

+1 for the former option.

> 
> Notes for some of the disabled checks:
> 
> TRY200, B904: Within an `except` clause, raise exceptions with `raise ... from err` or `raise from None. I've actually tried to implement [PEP-3134](https://peps.python.org/pep-3134/), as it looks like a good practice to me, but ultimately figured it's best to ask about it first.

I believe `raise ... from ...` is the current policy. tmt has a custom exception printer, exceptions logged will look different than what would the native interpreter logged, but it does support (and expect) exception chaining. While I did add `from ...` to all places I noticed, I did not touch *messages* of these exceptions - I agree `raise GeneralError(f"Invalid json syntax: {error}")` => `raise GeneralError(f"Invalid json syntax}") from error` is much better, it may result in a patch a bit more verbose, changing a lot of places. +1 from me, I'd just recommend making it a dedicated patch, or even a series of patches, e.g. one for each subpackage or something, if it turns out to be a very common issue. Keep @psss in the loop, BTW, he's often commenting on consistent logging and error messages, such a change fits the description pretty well :)

> Q: flake8-quotes (Q000 Single quotes found but double quotes preferred) would need changes in basically whole codebase:
> 
> > ruff . --select Q000
> > Found 4776 errors.

Yeah, probably not a good candidate for implementation... (Especially when I for one prefer single quotes :)

> C90: mccabe (complex structure)
Something like pylint, right? I'd leave this one for very long winter evenings, although it may serve as a pointer to way too complicated code, it might be cumbersome to get the threshold good enough.

> ANN: flake8-annotations - Some of the checks could compliment mypy.

I agree. I'm experimenting with `pyright`, to get another pair of eyes, but it does not play well with pre-commit...

> S: flake8-bandit - Security/vulnerability checks tmt probably doesn't need?

Depends. I'd give it a try, let's see what it reports, we can always disable some or even all checks.

> ISC: ISC001 and ISC003 is actually passing already. Not sure if/how to adapt the code for ISC002 (Implicitly concatenated string literals over multiple lines)

Hm, here I'm worried about readability. Might be better for our code to use multiline strings & `dedent()` instead of concatenating literals, but seems very much like a case-by-case area.

My thoughts on the rest of the commented-out checks:

* `"ARG",  # flake8-unused-arguments` - sounds good, but my experience tells me this particular check is mostly a waiver factory. I'd still give it a try before dropping it completely.
* `"BLE",  # flake8-blind-except` - I think we want this one
* `"FBT",  # flake8-boolean-trap` - no idea, let's try it out
* `"A",  # flake8-builtins` - let's try it out
* `"EM",  # flake8-errmsg` - seems to me dubious at best...
* `"G",  # flake8-logging-format` - tmt has its own logging methods, this might cause false positives, but let's give it a try - I see no use for `G001` or `G004` but `G201` or `G202` might be just an extra cheap bit of sanity checking
* `"T20",  # flake8-print` - I'd enable these too, but there will be code to fix for sure.
* `"SLF",  # flake8-self` - in the ideal world, yes, there will be violations as of now. I'd keep it around, and fix the reported issues.
* `"TCH",  # flake8-type-checking` - sounds like a good idea to try out, avoiding unnecessary imports, and keeping type-only imports in check...
* `"PLW",  # pylint-warning` - let's try it out, the list of checks looks solid
* `"TRY",  # tryceratops` - let's try it out, I have no idea what it might report, never heard of this linter, but checks look interesting

KwisatzHaderach commented on tmt/utils.py:
```diff
@@ -3044,15 +3049,15 @@ def _load(self, text: str) -> None:
         self.version(int(version_match.groups()[0]))
         log.debug(
-            "Detected StructuredField version {0}".format(
+            "Detected StructuredField version {}".format(
```
why are these not f-strings as all the others before?

apparently there are some more in convert, base and some here an there in other files as well

martinhoyer replied:
My guess is that f-strings are supposed to be always used if the variable(s) are already assigned, while `.format` is ok when we're doing some operation.

KwisatzHaderach replied:
don't know how others feel about this, but IMO it would be great if we had all of them as f-strings and ditch the str.format, on the other hand in some places it might look bad/weird

happz replied:
+1 for f-strings over `format()`. There's one exception, logging methods are allowed to use the printf-like formatting, `%d`, and so on, but my understanding was the f-strings were better & recommended everywhere else. And even if a computation is needed, it's either simple enough to fit into the f-string, or we can assign it to a helper variable first, for better readability, and use the f-string anyway.

martinhoyer replied:
Just to clarify, it's not like I'm against using f-strings everywhere, but there currently isn't any ruff rule/settings that would allow us to check for that, at least as far as I know.  
I'm happy to work on it and get rid of the remaining `format()`s, but maybe it would be more suitable to separate PR?

martinhoyer replied:
ok, I've added https://github.com/teemtee/tmt/pull/2014/commits/181d43eb09b165014f910a4bd453bef837e0df29 to hopefully get this resolved.  
I think it's not less (or more) readable than the str.format like this, though do agree helper variables could help with readability.

Tiboris commented on tmt/steps/provision/mrack.py:
```diff
@@ -528,25 +527,24 @@ def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecor
+                '--provision-tick',
+                metavar='SECONDS',
+                help=f'How often check Beaker for provisioning status, {DEFAULT_PROVISION_TICK} '
+                     f'seconds by default.'),
```
```suggestion
                     'seconds by default.'),
```

martinhoyer replied:
This took care of itself, when I was reverting the formatting changes :) Thanks though!

Tiboris commented on tmt/steps/provision/mrack.py:
```diff
@@ -528,25 +527,24 @@ def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecor
+                '--provision-timeout',
+                metavar='SECONDS',
+                help=f'How long to wait for provisioning to complete, {DEFAULT_PROVISION_TIMEOUT} '
+                     f'seconds by default.'),
```
```suggestion
                     'seconds by default.'),
```

Tiboris reviewed (commented):
I am sorry for doing a nitpick I have just realized an f-string is unnecessary on places without variables, see inline comments. If you could fix that for us 😇

martinhoyer commented on tmt/steps/execute/internal.py:
```diff
@@ -90,7 +95,7 @@ def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecor
             click.option(
                 '--no-progress-bar', is_flag=True,
                 help='Disable interactive progress bar showing the current test.')
```
aand I'm missing a comma. This is why you need trailing commas! :smiling_face_with_tear:

martinhoyer commented on tmt/utils.py:
```diff
@@ -4237,18 +4229,10 @@ def _load_keys(
-            logger.debug,
-            shift=LOG_SHIFT,
-            level=LOG_LEVEL,
-            topic=tmt.log.Topic.KEY_NORMALIZATION)
```
wow, how did that happened. Thanks, fixed.

martinhoyer commented on tmt/base.py:
```diff
@@ -2004,6 +2006,9 @@ def grow(
+            path = Path.cwd()
+
         return Tree(
             path=path,
```
Added `path=path or Path.cwd()` as suggested. Thanks!

martinhoyer commented:
``` ruff.....................................................................Failed
- hook id: ruff
- files were modified by this hook

Found 2 errors (2 fixed, 0 remaining).

All changes made by hooks:
diff --git a/tmt/utils.py b/tmt/utils.py
index 96f1c6a..ace05a7 100644
--- a/tmt/utils.py
+++ b/tmt/utils.py
@@ -1280,10 +1280,7 @@ def render_run_exception(exception: RunError) -> str:
     # Check verbosity level used during raising exception,
     # Supported way to correctly get verbosity is
     # tmt.util.Common.opt('verbose')
-    if isinstance(exception.caller, Common):
-        verbose = exception.caller.opt('verbose')
-    else:
-        verbose = 0
+    verbose = exception.caller.opt('verbose') if isinstance(exception.caller, Common) else 0
     for name, output in (('stdout', exception.stdout), ('stderr', exception.stderr)):
         if not output:
             continue
@@ -1295,10 +1292,7 @@ def render_run_exception(exception: RunError) -> str:
             line_summary = f"{min(len(output_lines), OUTPUT_LINES)}/{len(output_lines)}"
             output_lines = output_lines[-OUTPUT_LINES:]
 
-        lines += [
-            f'{name} ({line_summary} lines)',
-            OUTPUT_WIDTH * '~'
-            ] + output_lines + [
+        lines += [f'{name} ({line_summary} lines)', OUTPUT_WIDTH * '~', *output_lines] + [
```

huh, CI pre-commit failed,  but I'm not reproducing this locally with same version. That's fishy.
Will try to update pre-commit hook, which will require adding new `RUF008` and `RUF009` to ignores (or to fix those :) )

martinhoyer commented:
> > Notes for some of the disabled checks:
> > TRY200, B904: Within an `except` clause, raise exceptions with `raise ... from err` or `raise from None. I've actually tried to implement [PEP-3134](https://peps.python.org/pep-3134/), as it looks like a good practice to me, but ultimately figured it's best to ask about it first.
> 
> I believe `raise ... from ...` is the current policy. tmt has a custom exception printer, exceptions logged will look different than what would the native interpreter logged, but it does support (and expect) exception chaining. While I did add `from ...` to all places I noticed, I did not touch _messages_ of these exceptions - I agree `raise GeneralError(f"Invalid json syntax: {error}")` => `raise GeneralError(f"Invalid json syntax}") from error` is much better, it may result in a patch a bit more verbose, changing a lot of places. +1 from me, I'd just recommend making it a dedicated patch, or even a series of patches, e.g. one for each subpackage or something, if it turns out to be a very common issue. Keep @psss in the loop, BTW, he's often commenting on consistent logging and error messages, such a change fits the description pretty well :)

@psss, would you be up for enabling the B904 check and modifying the existing raises that does not use `raise from`?  
If yes, should it be part of this PR?  
Any guidelines for standardization of error messages/exception handling?

martinhoyer commented:
Rebased with main, updated ruff to 0.0.263 as it includes new checks and fix for SIM222.

martinhoyer commented:
Rebased with main.

happz commented:
/packit build

happz commented:
Let's be serious about this PR - added it to 1.24 O:) Maybe not everything would get in `main` in this time frame, but at least the PR would start among PRs the collective "we" would like to see happen, in some way or form.

martinhoyer commented:
> Let's be serious about this PR - added it to 1.24 O:) Maybe not everything would get in `main` in this time frame, but at least the PR would start among PRs the collective "we" would like to see happen, in some way or form.

Soo, why does it have a 'blocked' tag?

happz commented:
> > Let's be serious about this PR - added it to 1.24 O:) Maybe not everything would get in `main` in this time frame, but at least the PR would start among PRs the collective "we" would like to see happen, in some way or form.
> 
> Soo, why does it have a 'blocked' tag?

I felt it was blocked by some other PR, but that no longer applies.

martinhoyer commented:
> > > Let's be serious about this PR - added it to 1.24 O:) Maybe not everything would get in `main` in this time frame, but at least the PR would start among PRs the collective "we" would like to see happen, in some way or form.
> > 
> > 
> > Soo, why does it have a 'blocked' tag?
> 
> I felt it was blocked by some other PR, but that no longer applies.

ok, thanks. 
Rebased with main, updated pre-commit hook versions and added f-string conversion flags to pass RUF010.

lukaszachy commented:
/packit test

martinhoyer commented:
I've made some mistakes that caused the tests to fail, like not realizing some of the functions in test_utils were missing classmethod decorator.   

It's safe to say I've learned my lesson by opening a PR like this. Someone put me out of my misery :)

martinhoyer commented on tests/unit/test_cli.py:
```diff
@@ -231,7 +229,7 @@ class DecideColorizationTestcase:
 
 @pytest.mark.parametrize(
-    ('testcase',),
+    ('testcase',),  # noqa: PT006
```
No idea how to solve this.
PT006: `Wrong name(s) type in `@pytest.mark.parametrize`, expected `csv``

happz replied:
Hmm, no idea. A tuple of strings should be acceptable, or a coma-separated string, `"foo, bar"`.

happz replied:
I wouldn't mind, although I for one never used it, all pytest versions I have experience with were fine with `"foo, bar"` or `('foo', 'bar')`. I never had to name the first argument :)

martinhoyer replied:
ok, thanks. Using a string without argument names then.

happz commented:
/packit build

psss commented:
First, thanks much for taking time and doing such a significant code cleanup, @martinhoyer! Sorry, I got to it only today. It's a bit longer reading :-) From what I've seen, the changes look reasonable and beneficial but probably not a good idea to merge them now, right before the `1.24` release. What about trying to get them in early after the release?

> > > Notes for some of the disabled checks:
> > > TRY200, B904: Within an `except` clause, raise exceptions with `raise ... from err` or `raise from None. I've actually tried to implement [PEP-3134](https://peps.python.org/pep-3134/), as it looks like a good practice to me, but ultimately figured it's best to ask about it first.
> > 
> > 
> > I believe `raise ... from ...` is the current policy. tmt has a custom exception printer, exceptions logged will look different than what would the native interpreter logged, but it does support (and expect) exception chaining. While I did add `from ...` to all places I noticed, I did not touch _messages_ of these exceptions - I agree `raise GeneralError(f"Invalid json syntax: {error}")` => `raise GeneralError(f"Invalid json syntax}") from error` is much better, it may result in a patch a bit more verbose, changing a lot of places. +1 from me, I'd just recommend making it a dedicated patch, or even a series of patches, e.g. one for each subpackage or something, if it turns out to be a very common issue. Keep @psss in the loop, BTW, he's often commenting on consistent logging and error messages, such a change fits the description pretty well :)
> 
> @psss, would you be up for enabling the B904 check and modifying the existing raises that does not use `raise from`? If yes, should it be part of this PR? Any guidelines for standardization of error messages/exception handling?

The `B904` check sounds good, makes sense to me. I'd suggest to cover that in a separate pull request though, this one is already large enough.

martinhoyer commented:
@psss Thanks for checking it out! I shouldn't make it so large. Initially, I thought we could drop unwanted commits and/or cherry-pick good ones. I was stupid and have learned my lesson :)  

Next release sounds good. Just let me know when to do a rebase.

psss commented:
> Next release sounds good. Just let me know when to do a rebase.

I believe now would be the right time. Release is out, almost all approved pull requests merged. Let's fix existing conflicts, finish the review and merge soon. @martinhoyer, will you have some time this week?

martinhoyer commented:
> I believe now would be the right time. Release is out, almost all approved pull requests merged. Let's fix existing conflicts, finish the review and merge soon. @martinhoyer, will you have some time this week?

I hope so. Let me try to do a rebase today/tomorrow.

martinhoyer commented:
I'm not quite sure about how to deal with [RUF008](https://beta.ruff.rs/docs/rules/mutable-dataclass-default/) and [RUF009](https://beta.ruff.rs/docs/rules/function-call-in-dataclass-default-argument/).  
@happz You've been doing the dataclasses from what I can see - any ideas?

happz commented:
> I'm not quite sure about how to deal with [RUF008](https://beta.ruff.rs/docs/rules/mutable-dataclass-default/) and [RUF009](https://beta.ruff.rs/docs/rules/function-call-in-dataclass-default-argument/). @happz You've been doing the dataclasses from what I can see - any ideas?

* mutable-dataclass-default (RUF008) - yep, we don't want that, and even though some of our dataclasses do not use the `dataclasses.field()` anymore, this should be still helpful.
* function-call-in-dataclass-default-argument (RUF009) - sounds good as well, but, since a good portion of our dataclasses uses `tmt.utils.field()`, I guess this might be ridden with false positives. `mutable_default: list[int] = creating_list()` is bad, I agree with the example, but ours `some_key: List[str] = field(default_factory=list, normalize=normalize_string_list)` is perfectly common and valid in tmt code base. There does not seem to be any list of acceptable callables, or a "shadow `field`" variable, that would help. We can try to enable this one, but it's probably not going to be as useful as other checks.

martinhoyer commented:
>  mutable-dataclass-default (RUF008) - yep, we don't want that, and even though some of our dataclasses do not use the `dataclasses.field()` anymore, this should be still helpful.
>
Sorry, do you mean we don't want that check, or we don't want the existing code to remain as is?

> function-call-in-dataclass-default-argument (RUF009) - sounds good as well, but, since a good portion of our dataclasses uses `tmt.utils.field()`, I guess this might be ridden with false positives. `mutable_default: list[int] = creating_list()` is bad, I agree with the example, but ours `some_key: List[str] = field(default_factory=list, normalize=normalize_string_list)` is perfectly common and valid in tmt code base. There does not seem to be any list of acceptable callables, or a "shadow `field`" variable, that would help. We can try to enable this one, but it's probably not going to be as useful as other checks.

RUF009 takes [extend-immutable-calls](https://beta.ruff.rs/docs/settings/#flake8-bugbear-extend-immutable-calls) option. Perhaps we can use that?

happz commented:
> > mutable-dataclass-default (RUF008) - yep, we don't want that, and even though some of our dataclasses do not use the `dataclasses.field()` anymore, this should be still helpful.
> 
> Sorry, do you mean we don't want that check, or we don't want the existing code to remain as is?

Oh, sorry :)) Yes, the check looks like a good fit, we do not want mutable defaults of dataclass fields :)

> 
> > function-call-in-dataclass-default-argument (RUF009) - sounds good as well, but, since a good portion of our dataclasses uses `tmt.utils.field()`, I guess this might be ridden with false positives. `mutable_default: list[int] = creating_list()` is bad, I agree with the example, but ours `some_key: List[str] = field(default_factory=list, normalize=normalize_string_list)` is perfectly common and valid in tmt code base. There does not seem to be any list of acceptable callables, or a "shadow `field`" variable, that would help. We can try to enable this one, but it's probably not going to be as useful as other checks.
> 
> RUF009 takes [extend-immutable-calls](https://beta.ruff.rs/docs/settings/#flake8-bugbear-extend-immutable-calls) option. Perhaps we can use that?

Ah, I missed there are options. Right, this one looks like the exact control we'd like to extend with `tmt.utils.field`. That should make the check usable for tmt code.

So it's +1 & +1 :)

martinhoyer commented:
> So it's +1 & +1 :)

Any chance you could do the necessary changes please? :) @happz

happz commented:
> > So it's +1 & +1 :)
> 
> Any chance you could do the necessary changes please? :) @happz

Not in this PR, but I'd love to add one more once this one's sorted out, to enable the rules and make them pass. No problem.

happz commented:
> Looking at some of the test failures and scratching my head over how would the changes in this PR possibly caused them. Help? :)

* `/tests/libraries/local` - this seems to be unrelated to your patch, I've seen it failed elsewhere as well, https://github.com/teemtee/tmt/pull/2125#issuecomment-1589205355
* `/tests/multihost/*` and `/tests/report/*` - these refer to `multihost` prepare method, but that one has been recently removed from tmt code - I see it re-added here: https://github.com/teemtee/tmt/pull/2014/files#diff-2f45e2684afbb7caf4a24add9251b6ca1c4d678c04326cc830d7597d93d29f65R224

KwisatzHaderach commented:
The libraries bug should be fixed by https://github.com/teemtee/tmt/pull/2147

martinhoyer commented:
Thanks. Just the https://github.com/teemtee/tmt/pull/2147 remains then.

martinhoyer commented:
@psss rebased

happz commented on tests/unit/test_schemas.py:
```diff
@@ -81,7 +81,7 @@ def validate_node(tree, node, schema, label, name):
 """)
 
-        assert False, f'{label} {name} fails validation'
+        raise AssertionError(f"{label} {name} fails validation")
```
This one should probably be `pytest.fail` too.

martinhoyer replied:
Make sense, thanks.

happz commented on tmt/utils.py:
```diff
@@ -1420,13 +1420,9 @@ def render_run_exception(exception: RunError) -> List[str]:
-            ]
+        lines += [f'{name} ({line_summary} lines)',
+                  OUTPUT_WIDTH * '~',
+                  *output_lines, OUTPUT_WIDTH * '~', '']
```
Nitpicking, but: could you preserve the formatting? We're collecting a list of lines in `lines`, and `*output_lines, OUTPUT_WIDTH * `~`, ''` sort of hides it's actually N + 2 more lines by being on the same line of the code.

martinhoyer commented:
Resolved new conflicts with main, @happz's remarks and updated ruff version, as it fixed one noqa.

psss commented on tmt/base.py:
```diff
@@ -1339,7 +1339,7 @@ class Plan(
     _remote_plan_fmf_id: Optional[FmfId] = None
 
-    _extra_L2_keys = [
+    extra_l2_keys = [
```
Why is the underscore removed here? The list is not to be consumed by users outside of the `Plan` class.

martinhoyer replied:
my bad, I broke it while doing one of the many rebases.

psss commented on tmt/base.py:
```diff
@@ -324,7 +324,7 @@ class DependencyFmfId(FmfId):
     """
 
-    VALID_KEYS: ClassVar[List[str]] = FmfId.VALID_KEYS + ['destination', 'nick', 'type']
+    VALID_KEYS: ClassVar[List[str]] = [*FmfId.VALID_KEYS, 'destination', 'nick', 'type']
```
What's the point of these changes? Is this supposed to me more readable? I'd say the original is a bit more clear.

psss replied:
Ok, speed up. Makes sense.

psss commented on tmt/convert.py:
```diff
@@ -142,6 +140,7 @@ def html_to_markdown(html: str) -> str:
         markdown: str = ""
     else:
         markdown = md_handler.handle(html).strip()
+    # RET504: Unnecessary variable assignment before `return` statement. Keeping for readability.
```
Not needed?

martinhoyer replied:
yep, thanks.

psss reviewed (commented):
I've read through all the changes, looks good. Added just a few minor comments/questions.
